subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java

```diff
@@ -142,20 +142,53 @@ public Object transform(Object obj) {
         return Collections.emptyList();
     }
 
-    private static List<ExceptionPlaceholder> convertToExceptionPlaceholderList(List<? extends Throwable> throwables, Transformer<ExceptionReplacingObjectOutputStream, OutputStream> objectOutputStreamCreator) {
+    private static List<ExceptionPlaceholder> convertToExceptionPlaceholderList(Throwable original, List<? extends Throwable> throwables, Transformer<ExceptionReplacingObjectOutputStream, OutputStream> objectOutputStreamCreator, boolean checkCycle) {
         if (throwables.isEmpty()) {
             return Collections.emptyList();
         } else if (throwables.size() == 1) {
-            return Collections.singletonList(new ExceptionPlaceholder(throwables.get(0), objectOutputStreamCreator));
+            Throwable cause = throwables.get(0);
+            if (checkCycle) {
+                if (hasCycle(original, cause)) {
+                    return Collections.emptyList();
+                }
+            }
+            return Collections.singletonList(new ExceptionPlaceholder(cause, objectOutputStreamCreator));
         } else {
             List<ExceptionPlaceholder> placeholders = new ArrayList<ExceptionPlaceholder>(throwables.size());
             for (Throwable cause : throwables) {
-                placeholders.add(new ExceptionPlaceholder(cause, objectOutputStreamCreator));
+                if (!checkCycle || !hasCycle(original, cause)) {
+                    placeholders.add(new ExceptionPlaceholder(cause, objectOutputStreamCreator));
+                }
             }
             return placeholders;
         }
     }
 
+    private static boolean hasCycle(Throwable original, Throwable source) {
+        Throwable current;
+        Throwable cause = source;
+
+        try {
+            do {
+                current = cause;
+                if (original == current) {
+                    return true;
+                }
+                List<? extends Throwable> suppressed = extractSuppressed(current);
+                if (!suppressed.isEmpty()) {
+                    for (Throwable throwable : suppressed) {
+                        if (hasCycle(original, throwable)) {
+                            return true;
+                        }
+                    }
+                }
+            } while ((cause = current.getCause()) != null);
+        } catch (Throwable e) {
+            // Ignore, is handled elsewhere
```

Comment:
As discussed, if our lifecycle is fine with ignoring things like `ThreadDeath` or `InterruptedException`, this should be fine. Other than that, I wonder if we should actually ad a suppressed exception to `original` to indicate that we failed to handle this

Reply:
Refactoring the cycle check will make that block go away. As for doing something like that elsewhere in the class, I would wait until it is needed by a use case. That code has been "working" without it for a while.